docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,317 @@
+*   Discuss what modifications to a pod spec would be required for minimal
+    container troubleshooting
+
+Non-Goals of this proposal are:
+
+*   Fully design the ability to add a container to a pod spec of a running pod.
+    This is complex and should be covered in a separate doc if it's the agreed
+    upon solution.
+
+## Requirements
+
+A solution to troubleshoot minimal container images MUST:
+
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to all namespaces shared by containers in a pod and the mount
+    namespaces of individual containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
+    initialization
+*   respect admission restrictions
+*   allow introspection of pod state (no hidden containers)
+*   support arbitrary runtimes via the CRI (possibly with reduced feature set)
+
+A good solution SHOULD:
+
+*   have an excellent user experience (i.e. should be a feature of the platform
+    rather than a config-time solution)
+*   require no direct access to the node
+*   have no inherent side effects to the running container image
+*   re-use existing container image distribution channels
+*   enable detection of pods that have been modified by an exec ("tainted" pods)
+
+## Implementation
+
+To troubleshoot a single container, one executes a process inside the container
+namespaces using `kubectl exec`. Since a pod is a collection of containers, it
+follows that to debug a pod we could execute a container inside the pod
+namespaces. This could be a logical extension of `kubectl exec` by adding a new
+flag to specify the name of the container image to run inside the pod namespace.
+
+The user would specify a debug image of her choice (e.g. `kubectl exec -it -m
+gcr.io/myproj/debugimage podname`) and be connected to a process running inside
+the pod. The processes of all other containers would be visible through the
+shared PID namespace. Network connections are visible through the shared net
+namespace. File systems are available via volume bind mounts, potentially
+read-only.
+
+### Prerequisites
+
+We'll have to complete several independent prerequisites to realize this
+functionality, including a few long standing feature requests. These would be
+fully described in separate proposals.
+
+#### Shared PID Namespace
+
+Tracked in [#1615](https://issues.k8s.io/1615), a shared PID namespace became
+available in Docker 1.12 and requires plumbing through the docker drivers,
+extending the infra container to reap orphaned zombies, and [rolling
+out](https://pr.k8s.io/37404).
+
+#### Adding a Container to a Pod Spec
+
+Tracked in [#37838](https://issues.k8s.io/37838), Kubernetes disallows almost
+all modification to a pod spec as a conservative default and because arbitrary
+changes could have complex repercussions on resource accounting and quality of
+service. The most straightforward way to add a container to a running pod is to
+modify the pod spec.
```

Comment:
Primarily anything security related around pods, because their immutability today makes enforcement easier than otherwise.  I can think of 5-10 protections that depend on it.  I agree we will relax mutability, but at this point we need to carefully assess each change and make sure we're adding other controls to allow it (like field level auth)